### PR TITLE
Tiler module

### PR DESCRIPTION
**Pull request by lillythomas**
Tiling module that generates 256x256 tiled `xarray` datasets from the larger scene-level datacube generated in `datacube.py`.

To do:

- [x]  add cloud filter
- [x]  refine extent to tile within to valid data region
- [x] add main script to call the `datacube` module followed by the `tile` module
- [x] decide on method for storing the tiled `xarray` datasets
- [x] add some docstrings
- [x] plot tiles to make sure they look good

### CODE REVIEW

**Comment by yellowcap** [comment]
FYI I am using the following snippet to plot the merged data for visual cross-checks:
```python
def plot(merged):
    """
    Plot a tile in RGB and vh, and DEM
    """
    import numpy
    from matplotlib import pyplot as plt

    # data = merged.coarsen(x=100, y=100, boundary="trim").mean()
    data = merged

    fig, axs = plt.subplots(3)

    rgb = (
        numpy.array(
            [
                data.get("B04"),
                data.get("B03"),
                data.get("B02"),
            ]
        ).clip(0, 3000)
        / 3000
    )
    axs[0].imshow(rgb.transpose(1, 2, 0))
    axs[1].imshow(numpy.log(data.get("vh")))
    axs[2].imshow(data.get("DEM"))

    plt.show()
```

File: scripts/datacube.py [hunk 1]
```diff
@@ -397,24 +397,18 @@ def merge_datasets(ds_sen2, ds_sen1, da_dem):
...
+    # Merging the datasets as before
     ds_merge = xr.merge([ds_sen2, ds_sen1, da_dem], compat="override")
-    print("Merged dataset: ", ds_merge)
-    print(
-        "Time variables (S2, merged): ", ds_sen2.time.values, ds_merge.time.values
-    )  # ds_sen1.time.values, da_dem.time.values
-    return ds_merge
+
+    # Define a mask where all variables have valid data (not NaN)
+    valid_mask = ds_merge.notnull().all(dim=["x", "y"])
```
Comment (yellowcap) [inline scripts/datacube.py:406]: This is not necessary, the problem was with rounding errors on the WGS84 bbox. See https://github.com/Clay-foundation/model/pull/43

**Review by yellowcap — COMMENTED** [review]
Haven't had time to test this, was focused on fixing the cube. But looks good from a high level.

**Comment by lillythomas** [comment]
The `tiler` function seems to work nicely. Here is a plotted result `[RGB, VH, DEM]` in order from top to bottom. 

Screen Shot 2023-11-17 at 12 30 24 PM

**Comment by lillythomas** [comment]
Added a catch for sampled S2 reference dates which don't have S1 scenes within the +/- 3 day surrounding interval. This should prevent any failures due to unavailable matchup as it just triggers a new sampling for S2.

**Comment by lillythomas** [comment]
Going to merge this to keep things moving. Will address the tile saving method in a follow-up PR.

File: scripts/tile.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                    tuple(tile.dims[d] for d in ["x", "y"]),
+                    "; tile count: ",
+                    tile_count,
+                )
+                # Check for clouds and nodata
+                approval = filter_clouds_nodata(tile)
+                if approval == True:
+                    # Append the tile to the list
+                    tiles.append(tile)
+                else:
```
[the remark this answers is not in the document]
Comment (lillythomas) [reply scripts/tile.py]: Removed.

**Pull request merged**